### PR TITLE
[BEAM-4061] Introduced SpannerWriteResult

### PR DESCRIPTION
**Pull request by mairbek**
Introduced SpannerWriteResult that

- reports failed mutations;
- allows to sequence several writes to Cloud Spanner.

### CODE REVIEW

**Comment by mairbek** [comment]
R: @jkff

**Comment by mairbek** [comment]
Retest this please

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  public Map<TupleTag<?>, PValue> expand() {
+    return ImmutableMap.<TupleTag<?>, PValue>of(new TupleTag<Mutation>("failedMutation"),
+        failedMutations);
+  }
+
+  public PCollection<MutationGroup> getFailedMutations() {
+    return failedMutations;
+  }
+
+  public PCollection<Long> getFailedMutationsCount() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Hmm this lazy initialization seems a little funky. Is it necessary to have this as part of the SpannerWriteResult API? Seems like a user can trivially apply this themselves; especially considering that a simple Count.globally() is not sufficient in case they're writing an unbounded PCollection.

Btw can you please add a test with an unbounded PCollection? E.g. using TestStream, or using a simple Create.of() with .setBoundedInternal().

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (failedMutationsCount == null) {
+      failedMutationsCount = failedMutations.apply(Count.<MutationGroup>globally());
+    }
+    return failedMutationsCount;
+  }
+
+  @Override
+  public void finishSpecifyingOutput(String transformName, PInput input,
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Why is this needed, and why is doing nothing the right thing here?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 3]
```diff
@@ -939,16 +947,15 @@ public void processElement(ProcessContext c) throws Exception {
   }
 
   private static class WriteToSpannerFn
-      extends DoFn<Iterable<Mutation>, Void> {
-    private static final int MAX_RETRIES = 5;
-    private static final FluentBackoff BUNDLE_WRITE_BACKOFF = FluentBackoff.DEFAULT
-        .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+      extends DoFn<Iterable<MutationGroup>, MutationGroup> {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Could you please make this a DoFn returning some success indicator on the main output, and failed mutations on an additional output? That way this will be compatible with the Wait transform https://github.com/apache/beam/pull/4301.

Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: What do you mean by a success indicator? Isn't returning empty PCollection of failed mutations good enough?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 1 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: I assumed `Wait.on(x)` requires aggregated value, isn't that the case?

**Comment by reuvenlax** [comment]
any update on this PR?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 1 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: No, it doesn't require an aggregate. `a.apply(Wait.on(b))` means (roughly) "emit `a` after all elements of `b` with a lower timestamp".

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 3 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: I guess it's technically enough in this PR, but it can lead to issues in the future. E.g. in BigQueryIO, when using "batch loads", the collection of failed records is simply hard-coded as an empty collection, and it doesn't have a data dependency on writing successful records. If something like this is ever introduced to Spanner (e.g. bulk import), it can have a similar issue. I think, if a user wants to sequence something after a successful write, they should apply it to something that explicitly represents the results of successful writes. It can be something as simple as `Void` for now, or number of affected rows, or whatever.

**Comment by jkff** [comment]
Replied to comments, sorry for the delay - @mairbek PTAL.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 1 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Removed

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 2 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Doesn't compile otherwise

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 3 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: I see. I we return a number of applied mutations, but I'd have to count/aggregate that number. Won't that affect the write performance?

**Comment by mairbek** [comment]
Please take a look

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 4]
```diff
@@ -693,14 +701,18 @@ public Write withBatchSizeBytes(long batchSizeBytes) {
       return toBuilder().setBatchSizeBytes(batchSizeBytes).build();
     }
 
+    /** Specifies failure mode. */
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Document what is the default value

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 3 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Then I suggest you use `PCollection<Void>` or better `PCollection<?>` for now as the successful result type. Then you don't need to aggregate anything - but, if used as an input to Wait.on(), it will be "aggregated" very cheaply (all values ignored).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 5]
```diff
@@ -971,22 +977,27 @@ public void teardown() throws Exception {
...
+        } else {
+          throw new IllegalArgumentException("Unknown failure mode " + failureMode);
+        }
+      }
+      if (tryIndividual) {
+        for (MutationGroup mg : mutations) {
+          try {
+            spannerAccessor.getDatabaseClient().writeAtLeastOnce(mg);
+          } catch (SpannerException e) {
+            c.output(mg);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Would be good to add some logging to this function. If I got a bunch of failed mutations, I'd appreciate knowing why they failed. I don't know if it makes sense to embed the failure information into each failed mutation, it might.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 3 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Updated.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 5 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Done

**Comment by mairbek** [comment]
@chamikaramj @iemejia could you please take a look?

**Comment by iemejia** [comment]
Hi sorry I have missed the notification on this one. Does this one have a Jira if no can you please create one + add a description + rename PR title. Can you please also rebase and push to see the execution status.

**Comment by mairbek** [comment]
Done, @iemejia please take look when you have a chance

**Comment by chamikaramj** [comment]
What's the status of this ?

**Comment by iemejia** [comment]
HI, sorry I have forgotten about this one. I am not going to have a lot of availability next two weeks, so asking @jkff please to take this one back. Please excuse me @mairbek .

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  public PCollection<MutationGroup> getFailedMutations() {
+    return failedMutations;
+  }
+
+  public PCollection<Void> getOutput() {
+    return output;
+  }
+
+  @Override
+  public void finishSpecifyingOutput(String transformName, PInput input,
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Why this override?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  @Override
+  public Pipeline getPipeline() {
+    return pipeline;
+  }
+
+  @Override
+  public Map<TupleTag<?>, PValue> expand() {
+    return ImmutableMap.of(new TupleTag<Mutation>("failedMutation"), failedMutations);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: It seems safer if the TupleTag is passed in the constructor (can you use `failedTag`?), rather than re-created in expand()

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * A result of {@link SpannerIO#write()} transform. Use {@link #getFailedMutations} to access
+ * failed Mutations.
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Can you add a comment also about using `getOutput()` with the Wait transform?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java [hunk 9]
```diff
@@ -128,19 +140,72 @@ public void testWrite() throws Exception {
...
 
-    p.run();
-    DatabaseClient databaseClient =
-        spanner.getDatabaseClient(
-            DatabaseId.of(
-                project, options.getInstanceId(), databaseName));
+    PipelineResult result = p.run();
+    result.waitUntilFinish();
+    assertThat(result.getState(), is(PipelineResult.State.DONE));
+    assertThat(countNumberOfRecords(), equalTo((long) numRecords));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java]: Hm I wonder, can you use SpannerIO.read() here instead? (or readAll(), gated by Wait.on(write)) - that would be another cool demonstration of Wait.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 8 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java (hunk 7 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteResult.java]: Done

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java (hunk 9 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java]: Isn't it better for the test to compute that without going through beam? `testSequentialWrite` already tests `Wait.on`.

**Comment by jkff** [comment]
retest this please

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 10]
```diff
@@ -777,21 +797,23 @@ public PDone expand(PCollection<MutationGroup> input) {
...
-              "Write mutations to Spanner",
-              ParDo.of(new WriteToSpannerFn(spec.getSpannerConfig())));
-      return PDone.in(input.getPipeline());
-
+                  .withSideInputs(schemaView)).apply("Write mutations to Spanner",
+              ParDo.of(new WriteToSpannerFn(spec.getSpannerConfig(), spec.getFailureMode(),
+                  failedTag))
+                  .withOutputTags(mainTag, TupleTagList.of(failedTag)));
+      PCollection<MutationGroup> failedMutations = result.get(failedTag);
+      failedMutations.setCoder(SerializableCoder.of(MutationGroup.class));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: It's unfortunate that we can't directly use MutationGroupEncoder as a Coder, since it requires a Schema known at construction time...

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java (hunk 9 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteIT.java]: In general, "write-then-read" tests are preferred in Beam, where both write and read are done using your transform. They can be done in separate pipelines in one test method, or in one pipeline. I'm OK keeping this as-is for this PR though.

**Comment by jkff** [comment]
There seems to be a checkstyle violation that's legit, could you fix it? But also a bunch of Java compiler OOMs, not sure why is that, will ping dev@.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 10 above)
Comment (mairbek) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java]: Agree, if you have ideas how to improve this I'll be happy to help

**Comment by mairbek** [comment]
Fixed checkstyle warnings

**Comment by jkff** [comment]
Looks legit, but let's retest one more time...

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
OK, the test failure is flaky, and I confirmed locally that the tests pass. Merging.
